hwaddress-sanitizer/run_in_qemu_with_lam.sh

```diff
@@ -0,0 +1,158 @@
+#!/bin/bash -eu
+#
+# Runs the given binary inside QEMU, with the Intel LAM feature enabled.  Uses
+# full system mode due to issues with sanitizers in user mode.
+#
+# Usage:  ./run_in_qemu_with_lam.sh binary-to-run [argument]...
+#
+# NOTE:  You also need to ensure the environment variables below point to the
+# necessary prerequisites: qemu-system-x86_64, qemu-img, debian-10.qcow2,
+# debian-10.key, bzImage, and llvm-symbolizer.
+
+# Path to binary to run in QEMU.
+readonly BINARY_PATH="${1}"
+
+# Arguments to pass the binary inside QEMU.
+readonly BINARY_ARGS="${@:2}"
+
+# Path to a qemu-system-x86_64 binary built with LAM support.
+: ${QEMU:="${PWD}/qemu/build/qemu-system-x86_64"}
+
+# Path to a qemu-img binary.
+: ${QEMU_IMG:="${PWD}/qemu/build/qemu-img"}
+
+# Path to a Debian 10 image configured with auto-login and SSH.
+: ${IMAGE:="${PWD}/debian-10.qcow2"}
+
+# Path to the SSH key for the root user of the Debian image.
+: ${SSH_KEY:="${PWD}/debian-10.key"}
+
+# The directory inside the Debian image where tests should run.  Must exist
+# prior to running this script.
+: ${QEMU_WORKSPACE_PATH:="/workspace"}
+
+# Path to a Linux kernel bzImage built with LAM support.
+: ${KERNEL:="${PWD}/linux/build/arch/x86_64/boot/bzImage"}
+
+# Path to an llvm-symbolizer built with the following config:
+#   cmake -GNinja -DLLVM_BUILD_RUNTIME=OFF -DCMAKE_BUILD_TYPE=Release \
+#       -DLLVM_STATIC_LINK_CXX_STDLIB=ON ../llvm/
+: ${LLVM_SYMBOLIZER:="${PWD}/llvm-project/build/bin/llvm-symbolizer"}
+
+: ${HWASAN_OPTIONS:=""}
+
+readonly QEMU_FORCE_KILL_TIMEOUT=3
+readonly QEMU_EXIT_STATUS_PREFIX="command_exit_status:"
+readonly HOST_TMPDIR="$(mktemp -d)"
+readonly DELTA_IMAGE="${HOST_TMPDIR}/delta.img"
+readonly QEMU_PIPE="${HOST_TMPDIR}/qemu"
+readonly QEMU_INPUT_PIPE="${QEMU_PIPE}.in"
+readonly QEMU_OUTPUT_PIPE="${QEMU_PIPE}.out"
+readonly BINARY_NAME="$(basename ${BINARY_PATH})"
+
+QEMU_PID=""
+SSH_PORT=""
+
+function force_kill_qemu_after_timeout {
+  sleep "${QEMU_FORCE_KILL_TIMEOUT}"
+  kill -9 "${QEMU_PID}" &>/dev/null || true
+}
+
+function on_exit {
+  (force_kill_qemu_after_timeout) &
+  if kill "${QEMU_PID}"; then
+    echo "Waiting for QEMU to shutdown..."
+    wait "${QEMU_PID}" &>/dev/null || true
+  fi
+
+  echo "Done!"
+}
+
+function wait_pipe_for_match {
```

Comment:
why pipes and not just ssh \<command\>?
If this is to keep just a single ssh connection then "-M" should be enough

Reply:
I still don't fully understand what's going on, but it seems that having too many SSH commands running in parallel is triggering the failures.  If I reduce the number of llvm-lit workers, I get fewer failures.

I don't have this issue using the pipe method implemented here.

Reply:
Even with -M? For -M you need to keep the first ssh connection open.

Reply:
Thanks.  Indeed `-M` seems to be the trick.  Using it, I actually get faster test executions than using pipes, *and* the strange test failures go away.

I've uploaded a new diff with this approach.